src/analyses/variable-sensitivity/full_struct_abstract_object.cpp

```diff
@@ -182,29 +194,31 @@ sharing_ptrt<struct_abstract_objectt>
         member_expr.compound().type(), false, true));
   }
 
-  internal_sharing_ptrt<full_struct_abstract_objectt> copy(
-    new full_struct_abstract_objectt(*this));
+  const auto &result=
+    std::dynamic_pointer_cast<full_struct_abstract_objectt>(mutable_clone());
 
   if(!stack.empty())
   {
     abstract_object_pointert starting_value;
-    irep_idt c=member_expr.get_component_name();
-    if(map.find(c)==map.cend())
+    const irep_idt c=member_expr.get_component_name();
+    shared_struct_mapt::const_find_type old_value = map.find(c);
+    if(!old_value.second)
     {
       starting_value=
         environment.abstract_object_factory(
           member_expr.type(), ns, true, false);
     }
     else
     {
-      starting_value=map.at(c);
+      starting_value=old_value.first;
     }
 
-    copy->map[c]=
+    result->map[c] =
```

Comment:
Here (and in lines 252 and 256 below) we could first check whether the new value we're about to write to the map is already the same one that is in the map. Then we could avoid breaking the sharing in these cases.

To write an item to the map, and we know that the key is in the map (because we've already checked it earlier), we can use the non-const `find(..., tvt(true)` instead of `[]` to retrieve a reference. The second parameter tells it that the key is definitely in the map, in which case it needs to traverse the path from the root to the item only once (otherwise it does two passes, one read-only pass to determine whether the key is in the map, and if it is, it does a second pass in which it copies the nodes on the path from the root to the item).

To put an element into the map and we know that it is not in it, we can use `insert(..., tvt(false))` which again has the effect that the path from the root to the element is followed only once.